src/iceberg/avro/avro_direct_decoder.cc

```diff
@@ -0,0 +1,597 @@
+      decoder.skipString();
+      return {};
+
+    case ::avro::AVRO_BYTES:
+      decoder.skipBytes();
+      return {};
+
+    case ::avro::AVRO_FIXED:
+      decoder.skipFixed(avro_node->fixedSize());
+      return {};
+
+    case ::avro::AVRO_RECORD: {
+      // Skip all fields in order
+      for (size_t i = 0; i < avro_node->leaves(); ++i) {
+        ICEBERG_RETURN_UNEXPECTED(SkipAvroValue(avro_node->leafAt(i), decoder));
+      }
+      return {};
+    }
+
+    case ::avro::AVRO_ENUM:
+      decoder.decodeEnum();
+      return {};
+
+    case ::avro::AVRO_ARRAY: {
+      const auto& element_node = avro_node->leafAt(0);
+      // skipArray() returns count like arrayStart(), must handle all blocks
+      int64_t block_count = decoder.skipArray();
+      while (block_count > 0) {
+        for (int64_t i = 0; i < block_count; ++i) {
+          ICEBERG_RETURN_UNEXPECTED(SkipAvroValue(element_node, decoder));
+        }
+        block_count = decoder.arrayNext();
+      }
+      return {};
+    }
+
+    case ::avro::AVRO_MAP: {
+      const auto& value_node = avro_node->leafAt(1);
+      // skipMap() returns count like mapStart(), must handle all blocks
+      int64_t block_count = decoder.skipMap();
+      while (block_count > 0) {
+        for (int64_t i = 0; i < block_count; ++i) {
+          decoder.skipString();  // Skip key (always string in Avro maps)
+          ICEBERG_RETURN_UNEXPECTED(SkipAvroValue(value_node, decoder));
+        }
+        block_count = decoder.mapNext();
+      }
+      return {};
+    }
+
+    case ::avro::AVRO_UNION: {
+      const size_t branch_index = decoder.decodeUnionIndex();
+      // Validate branch index
+      const size_t num_branches = avro_node->leaves();
+      if (branch_index >= num_branches) {
+        return InvalidArgument("Union branch index {} out of range [0, {})", branch_index,
+                               num_branches);
+      }
+      return SkipAvroValue(avro_node->leafAt(branch_index), decoder);
+    }
+
+    default:
+      return InvalidArgument("Unsupported Avro type for skipping: {}",
+                             ToString(avro_node));
+  }
+}
+
+/// \brief Decode Avro record directly to Arrow struct builder.
+Status DecodeStructToBuilder(const ::avro::NodePtr& avro_node, ::avro::Decoder& decoder,
+                             const std::span<const FieldProjection>& projections,
+                             const StructType& struct_type,
+                             ::arrow::ArrayBuilder* array_builder) {
+  if (avro_node->type() != ::avro::AVRO_RECORD) {
+    return InvalidArgument("Expected Avro record, got type: {}", ToString(avro_node));
+  }
+
+  auto* struct_builder = internal::checked_cast<::arrow::StructBuilder*>(array_builder);
+  ICEBERG_ARROW_RETURN_NOT_OK(struct_builder->Append());
+
+  // Build a map from Avro field index to projection index
+  // -1 means the field should be skipped
+  std::vector<int> avro_to_projection(avro_node->leaves(), -1);
```

Comment:
I'm not sure if we need to cache `avro_to_projection` somewhere, perhaps in the `DecodeContext` that I propose in the below comment. We can use `const FieldProjection*` as the key to find the cached value.